### PR TITLE
Translate closures through the collector

### PR DESCRIPTION
**Pull request by arielb1**
Now that old trans is gone, there is no need for the hack of translating closures when they are instantiated. We can translate them like regular items.

r? @eddyb

### CODE REVIEW

File: src/librustc/ty/sty.rs [hunk 1]
```diff
@@ -254,15 +257,38 @@ pub enum TypeVariants<'tcx> {
...
+impl<'tcx> FnMut<(&'tcx Kind<'tcx>,)> for KindAsType {
+    extern "rust-call" fn call_mut(&mut self, args: (&'tcx Kind<'tcx>,)) -> Ty<'tcx> {
+        args.0.as_type().expect("unexpected region in upvars")
+    }
+}
+
+impl<'a, 'gcx, 'acx, 'tcx> ClosureSubsts<'tcx> {
+    #[inline]
+    pub fn upvar_tys(self, def_id: DefId, tcx: TyCtxt<'a, 'gcx, 'acx>) ->
+        iter::Map<slice::Iter<'tcx, Kind<'tcx>>, KindAsType>
```
Comment (eddyb) [inline src/librustc/ty/sty.rs]: Should just use `impl Trait` here, maybe you can share the error?
Last I checked most cases can be tricked into inferring correctly.

File: src/librustc/ty/sty.rs [hunk 2]
```diff
@@ -254,15 +257,38 @@ pub enum TypeVariants<'tcx> {
 /// handle). Plus it fixes an ICE. :P
 #[derive(Copy, Clone, PartialEq, Eq, Hash, Debug, RustcEncodable, RustcDecodable)]
 pub struct ClosureSubsts<'tcx> {
-    /// Lifetime and type parameters from the enclosing function.
+    /// Lifetime and type parameters from the enclosing function,
+    /// concatenated with the types of the upvars.
```
Comment (eddyb) [inline src/librustc/ty/sty.rs]: This is pretty much my only concern, that the upvar types are here - in fact, why are they in the type at all?
Why can't they be like ADTs, with the closure pointing to a list of "fields" which are inferred by typeck?

Comment (arielb1) [inline src/librustc/ty/sty.rs]: Before trans, closure types need the upvars because they contain lifetimes.

File: src/librustc/ty/sty.rs (hunk 1 above)
Comment (arielb1) [inline src/librustc/ty/sty.rs]:
```
error[E0564]: only named lifetimes are allowed in `impl Trait`, but `` was found in the type `core::iter::Map<core::slice::Iter<'_, ty::subst::Kind<'tcx>>, [closure@../src/librustc/ty/sty.rs:292:13: 292:66]>`
```

File: src/librustc/ty/sty.rs (hunk 2 above)
Comment (eddyb) [inline src/librustc/ty/sty.rs]: But couldn't the types with those lifetimes be stored *somewhere else*, keyed by `DefId`?
I believe we already do something like that for the signature of the closure.

**Comment by eddyb** [comment]
cc @rust-lang/compiler

File: src/librustc/ty/sty.rs (hunk 1 above)
Comment (eddyb) [inline src/librustc/ty/sty.rs]: Try `+ 'tcx` on the `impl Iterator`.

**Comment by michaelwoerister** [comment]
Nice `:)`
One thing to note is that after these changes, closures are only instantiated if they are called or there's a vtable for them. Before, we instantiated closures as soon as they were declared. For incremental compilation, the old behavior would provide more stability but in practice this probably won't come up often (or maybe not at all).

File: src/librustc_typeck/check/closure.rs [hunk 3]
```diff
@@ -40,19 +45,62 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
+
+            tcx.alloc_generics(ty::Generics {
+                parent: Some(base_def_id),
+                parent_regions: base_generics.parent_regions +
+                    (base_generics.regions.len() as u32),
+                parent_types: base_generics.parent_types +
+                    (base_generics.types.len() as u32),
+                regions: vec![],
+                types: upvar_decls,
+                has_self: false,
```
Comment (eddyb) [inline src/librustc_typeck/check/closure.rs]: This should be `base_generics.has_self`. Nevertheless, putting generics/type *here* is a problem.
In #37676 I've [added generics for closures in `typeck::collect`](https://github.com/rust-lang/rust/pull/37676/commits/68c898bfc2c1efdcfd1b0eabdf8217415a7af77e#diff-4ed25c00aceb84666fca639cf8101c7cR137), which is closer to what on-demand type-checking needs.
OTOH your `closure_base_def_id` is better (or even actually correct) than what I have there, so I'd prefer it.

**Comment by eddyb** [comment]
@arielb1 r=me with the closure generics/type creation moved to `typeck::collect` as per my comment.

**Pull request merged**